### DISCUSSION TITLE
Shared storage for multi node training

### DISCUSSION THREAD

**Discussion by roywei** [skypilot-org/skypilot#1845] [General]
Hi, 

I love this project and great work! One question I have is what's the file storage size limit and throughput limit for a shared file system that all nodes can access (read/write)? It's not very clear from the [storage doc](https://skypilot.readthedocs.io/en/latest/reference/storage.html), and what's the underlying implementation?

Thanks and can't wait to try this out!

**Comment by romilbhardwaj** [comment]
Hi @roywei, thanks for your interest!

The shared FS example in the docs uses an object store (S3/GCS/R2) as the backing persistent store. We provide access to these object stores by creating a FUSE mount using either [goofys](https://github.com/kahing/goofys) (for S3 and R2) or [gcsfuse](https://github.com/GoogleCloudPlatform/gcsfuse) (for gcs).

Performance varies depending on your chosen VM type, the backing store and the workload. For S3, we ran some [benchmarks](https://github.com/skypilot-org/skypilot/blob/master/examples/perf/storage_rawperf.yaml) and the detailed results can be found [here](https://github.com/skypilot-org/skypilot/blob/master/examples/perf/results.md). To summarize, on AWS m5.8xlarge with default EBS volume attached with 4 parallel processes reading/writing to files, we observed:

### Disk Bandwidth (MB/s)

|                  | EBS | SkyPilot Storage<br/>(S3, MOUNT mode) |
|------------------|-----|----------------------------------|
| Sequential Read  | 130 | 642                              |
| Sequential Write | 129 | 1828                             |


### Disk IOPS

|                  | EBS  | SkyPilot Storage<br/>(S3, MOUNT mode) |
|------------------|------|----------------------------------|
| Sequential Read  | 2051 | 8462                             |
| Sequential Write | 2055 | 27899                            | 

Please note that performance is highly dependent on the workload. The discussion section in the [results](https://github.com/skypilot-org/skypilot/blob/master/examples/perf/results.md) file provides cases when shared storage performance might be low (e.g., many small files).

Maximum file size, though I haven't tested it, should be similar to the limits put in by the backing object store. [Here's](https://aws.amazon.com/s3/faqs/#:~:text=Individual%20Amazon%20S3%20objects%20can,using%20the%20multipart%20upload%20capability.) S3's limits.

If you have a particular use case in mind, I am happy to discuss it!

**Reply by roywei** [reply]
Thanks a lot for the pointers! This is really helpful. I'm looking for how scalable is the shared storage and the perf numbers look decent. Technically it's also not hard to support lustre based high performance file system for really large scale. References in [AWS](https://docs.aws.amazon.com/fsx/latest/LustreGuide/what-is.html), [GCP](https://cloud.google.com/architecture/lustre-architecture), and [Azure](https://techcommunity.microsoft.com/t5/azure-high-performance-computing/lustre-on-azure/ba-p/1052536)

**Reply by romilbhardwaj** [reply]
Just to add, using object stores + fuse mounting has its fair share of problems we run into:

1. Poor performance when using many small files because of increased S3 API calls
2. Lack of support for all POSIX operations (e.g., appends are not supported)

Lustre would indeed be awesome to support as a high performance FS option. Main requirements would be 1) should expose a POSIX interface (i.e., should be mountable so users don't have to change code) and 2) should be accessible across regions, in case a task gets moved. Happy to collaborate on this if it is of interest to you!

**Reply by roywei** [reply]
Sure, will create a RFC when I need this (no timeline promised :)). [AWS file cache](https://docs.aws.amazon.com/fsx/latest/FileCacheGuide/what-is.html) supports multi region access, but it's really expensive.

**Reply by romilbhardwaj** [reply]
Great, looking forward to it! Meanwhile if you run into any issues with SkyPilot, please let us know!
